README.md
[![Actions Status](https://github.com/sanko/LibUI.pm/actions/workflows/linux.yaml/badge.svg)](https://github.com/sanko/LibUI.pm/actions) [![Actions Status](https://github.com/sanko/LibUI.pm/actions/workflows/windows.yaml/badge.svg)](https://github.com/sanko/LibUI.pm/actions) [![Actions Status](https://github.com/sanko/LibUI.pm/actions/workflows/osx.yaml/badge.svg)](https://github.com/sanko/LibUI.pm/actions) [![Actions Status](https://github.com/sanko/LibUI.pm/actions/workflows/freebsd.yaml/badge.svg)](https://github.com/sanko/LibUI.pm/actions) [![MetaCPAN Release](https://badge.fury.io/pl/LibUI.svg)](https://metacpan.org/release/LibUI)
# NAME

LibUI - Simple, Portable, Native GUI Library

# SYNOPSIS

    use LibUI ':all';
    use LibUI::Window;
    use LibUI::Label;
    Init( ) && die;
    my $window = LibUI::Window->new( 'Hi', 320, 100, 0 );
    $window->setMargined( 1 );
    $window->setChild( LibUI::Label->new('Hello, World!') );
    $window->onClosing(
        sub {
            Quit();
            return 1;
        },
        undef
    );
    $window->show;
    Main();

<div>
    <h2>Screenshots</h2> <div style="text-align: center">
    <h3>Linux</h3><img alt="Linux"
    src="https://sankorobinson.com/LibUI.pm/screenshots/synopsis/linux.png" />
    <h3>MacOS</h3><img
    alt="MacOS"
    src="https://sankorobinson.com/LibUI.pm/screenshots/synopsis/macos.png" />
    <h3>Windows</h3><img
    alt="Windows"
    src="https://sankorobinson.com/LibUI.pm/screenshots/synopsis/windows.png" />
    </div>
</div>

# DESCRIPTION

LibUI is a simple and portable (but not inflexible) GUI library in C that uses
the native GUI technologies of each platform it supports.

This distribution is under construction. It works but is incomplete.

# Container controls

- [LibUI::Window](https://metacpan.org/pod/LibUI%3A%3AWindow) - a top-level window
- [LibUI::HBox](https://metacpan.org/pod/LibUI%3A%3AHBox) - a horizontally aligned, boxlike container that holds a group of controls
- [LibUI::VBox](https://metacpan.org/pod/LibUI%3A%3AVBox) - a vertically aligned, boxlike container that holds a group of controls
- [LibUI::Tab](https://metacpan.org/pod/LibUI%3A%3ATab) - a multi-page control interface that displays one page at a time
- [LibUI::Group](https://metacpan.org/pod/LibUI%3A%3AGroup) - a container that adds a label to the child
- [LibUI::Form](https://metacpan.org/pod/LibUI%3A%3AForm) - a container to organize controls as labeled fields
- [LibUI::Grid](https://metacpan.org/pod/LibUI%3A%3AGrid) - a container to arrange controls in a grid

# Data entry controls

- [LibUI::Button](https://metacpan.org/pod/LibUI%3A%3AButton) - a button control that triggers a callback when clicked
- [LibUI::Checkbox](https://metacpan.org/pod/LibUI%3A%3ACheckbox) - a user checkable box accompanied by a text label
- [LibUI::Entry](https://metacpan.org/pod/LibUI%3A%3AEntry) - a single line text entry field
- [LibUI::PasswordEntry](https://metacpan.org/pod/LibUI%3A%3APasswordEntry) - a single line, obscured text entry field
- [LibUI::SearchEntry](https://metacpan.org/pod/LibUI%3A%3ASearchEntry) - a single line search query field
- [LibUI::Spinbox](https://metacpan.org/pod/LibUI%3A%3ASpinbox) - display and modify integer values via a text field or +/- buttons
- [LibUI::Slider](https://metacpan.org/pod/LibUI%3A%3ASlider) - display and modify integer values via a draggable slider
- [LibUI::Combobox](https://metacpan.org/pod/LibUI%3A%3ACombobox) - a drop down menu to select one of a predefined list of items
- [LibUI::EditableCombobox](https://metacpan.org/pod/LibUI%3A%3AEditableCombobox) - a drop down menu to select one of a predefined list of items or enter you own
- [LibUI::RadioButtons](https://metacpan.org/pod/LibUI%3A%3ARadioButtons) - a multiple choice control of check buttons from which only one can be selected at a time
- [LibUI::DateTimePicker](https://metacpan.org/pod/LibUI%3A%3ADateTimePicker) - a control to enter a date and/or time
- [LibUI::DatePicker](https://metacpan.org/pod/LibUI%3A%3ADatePicker) - a control to enter a date
- [LibUI::TimePicker](https://metacpan.org/pod/LibUI%3A%3ATimePicker) - a control to enter a time
- [LibUI::MultilineEntry](https://metacpan.org/pod/LibUI%3A%3AMultilineEntry) - a multi line entry that visually wraps text when lines overflow
- [LibUI::NonWrappingMultilineEntry](https://metacpan.org/pod/LibUI%3A%3ANonWrappingMultilineEntry) - a multi line entry that scrolls text horizontally when lines overflow
- [LibUI::FontButton](https://metacpan.org/pod/LibUI%3A%3AFontButton) - a control that opens a font picker when clicked

## Static controls

- [LibUI::Label](https://metacpan.org/pod/LibUI%3A%3ALabel) - a control to display non-interactive text
- [LibUI::ProgressBar](https://metacpan.org/pod/LibUI%3A%3AProgressBar) - a control that visualizes the progress of a task via the fill level of a horizontal bar
- [LibUI::HSeparator](https://metacpan.org/pod/LibUI%3A%3AHSeparator) - a control to visually separate controls horizontally
- [LibUI::VSeparator](https://metacpan.org/pod/LibUI%3A%3AVSeparator) - a control to visually separate controls vertically

## Dialog windows

- `openFile( )` - File chooser to select a single file
- `openFolder( )` - File chooser to select a single folder
- `saveFile( )` - Save file dialog
- `msgBox( ... )` - Message box dialog
- `msgBoxError( ... )` - Error message box dialog

See ["Dialog windows" in LibUI::Window](https://metacpan.org/pod/LibUI%3A%3AWindow#Dialog-windows) for more.

## Menus

- `LibUI::Menu` - application-level menu bar
- `LibUI::MenuItem` - menu items used in conjunction with [LibUI::Menu](https://metacpan.org/pod/LibUI%3A%3AMenu)

## Tables

The upstream API is a mess so I'm still plotting around this.

# GUI Functions

Some basics you gotta use just to keep a modern GUI running.

This is incomplete but... well, I'm working on it.

## `Init( [...] )`

    Init( );

Ask LibUI to do all the platform specific work to get up and running. If LibUI
fails to initialize itself, this will return a true value. Weird upstream
choice, I know...

You **must** call this before creating widgets.

## `Main( ... )`

    Main( );

Let LibUI's event loop run until interrupted.

## `Uninit( ... )`

    Uninit( );

Ask LibUI to break everything down before quitting.

## `Quit( ... )`

    Quit( );

Quit.

## `Timer( ... )`

    Timer( 1000, sub { die 'do not do this here' }, undef);

    Timer(
        1000,
        sub {
            my $data = shift;
            return 1 unless ++$data->{ticks} == 5;
            0;
        },
        { ticks => 0 }
    );

Expected parameters include:

- `$time`

    Time in milliseconds.

- `$func`

    CodeRef that will be triggered when `$time` runs out.

    Return a true value from your `$func` to make your timer repeating.

- `$data`

    Any userdata you feel like passing. It'll be handed off to your function.

# Requirements

See [Alien::libui](https://metacpan.org/pod/Alien%3A%3Alibui)

# See Also

`eg/demo.pl` - Very basic example

`eg/widgets.pl` - Demo of basic controls

# LICENSE

Copyright (C) Sanko Robinson.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

# AUTHOR

Sanko Robinson <sanko@cpan.org>


Changes
Revision history for Perl extension LibUI.pm

{{$NEXT}}

    - Sugary use statement

0.02 2023-03-11T05:00:07Z

    - New widgets:
        - LibUI::Area (very rough)
        - LibUI::ScrollingArea
    - Renamed widgets to fit existing pattern:
        - LibUI::HorizontalSeparator => LibUI::HSeparator
        - LibUI::VerticalSeparator   => LibUI::VSeparator
    - New example scripts:
        - eg/demo.pl: very basic, "Hello, World!" type of example
        - eg/widgets.pl: demo of (nearly all) basic controls
        - eg/histogram.pl: demo of LibUI::Area and releated functions
    - New enum and structs: LibUI::Area::Handler, ::KeyEvent, ::MouseEvent, ::Modifiers

0.01 2022-12-07T00:34:14Z

    - original version


lib/LibUI.pm
use 5.008001;

package LibUI 0.03 {
    use strict;
    use warnings;
    use lib '../lib', '../blib/arch', '../blib/lib';
    use Affix;
    use Alien::libui;
    use parent 'Exporter';    # gives you Exporter's import() method directly
    use Config;
    use Carp;
    our %EXPORT_TAGS;
    $|++;
    #
    #my $path = '/home/sanko/Downloads/libui-ng-master/build/meson-out/libui.so.0';
    sub lib () { CORE::state $lib //= Alien::libui->dynamic_libs; $lib }
    #
    my @exports;

    sub import {
        my ( $pkg, @args ) = @_;
        my $callpkg = caller(0);
        for my $arg (@args) {
            if ( $arg =~ m[^::(.+)$] ) {
                my $widget = 'LibUI' . $arg;
                eval qq{use $widget};
                croak $@ if $@;
                {
                    my $hash = $widget::{EXPORT_TAGS};
                    for my $key ( keys %$hash ) {
                        warn $key;
                        if ( exists $EXPORT_TAGS{$key} ) {
                            push @{ $EXPORT_TAGS{$key} }, $hash->{$key};
                        }
                        else {
                            $EXPORT_TAGS{$key} = $hash->{$key};
                        }
                    }
                }
            }
            elsif ( $arg =~ m[^:(.+)$] ) {
                push @exports, @{ $EXPORT_TAGS{$1} };
            }
        }
        no strict 'refs';
        *{"$callpkg\::$_"} = \&{"$pkg\::$_"} for @exports;
    }

    # TODO: unimport
    sub export {
        my ( $tag, @funcs ) = @_;
        push @{ $EXPORT_TAGS{$tag} }, map { m[^ui]; $'; } @funcs;
    }

    sub func {
        my ( $func, $params, $ret ) = @_;
        my $name = $func;
        $name =~ s[^ui][LibUI::];
        $name
            =~ s[LibUI::(Box|Button|Combobox|Control|Menu|NonWrappingMultilineEntry|RadioButtons|Slider|Window)][LibUI::$1::];
        $name =~ s[::New(.+)$][::$1::new];

        #warn sprintf '%30s => %-50s', $func, $name;
        affix( lib, [ $func, $name ], $params, $ret );
    }
    #
    sub Init {
        my $aggs = @_ ? shift : { Size => 1024 };
        CORE::state $func
            //= wrap( lib(), 'uiInit', [ Pointer [ Struct [ Size => Size_t ] ] ], Str );
        $func->($aggs);
    }

    sub Timer($&;$) {
        my ( $timeout, $coderef, $agg ) = @_;
        CORE::state $func
            //= wrap( lib(), 'uiTimer', [ Int, CodeRef [ [Any] => Int ], Any ] => Void );
        $func->( $timeout, $coderef, $agg // undef );
    }
    typedef uiForEach => Enum [qw[uiForEachContinue uiForEachStop]];
    {
        export default => qw[uiInit uiUninit uiFreeInitError
            uiMain uiMainSteps uiMainStep uiQuit
            uiQueueMain
            uiTimer
            uiFreeText
        ];
        func( 'uiUninit', [] => Void );
        func( 'uiMain',   [] => Void );
        func( 'uiQuit',   [] => Void );

        # Undocumented
        func( 'uiFreeInitError', [Str] => Void );
        #
        func( 'uiMainSteps', []    => Void );
        func( 'uiMainStep',  [Int] => Int );
        #
        func( 'uiQueueMain', [ CodeRef [ [ Pointer [Void] ] => Void ], Any ] => Void );
        #
        affix(
            lib(),
            [ 'uiOnShouldQuit',         'LibUI::onShouldQuit' ],
            [ CodeRef [ [Any] => Int ], Any ] => Void
        );
        func( 'uiFreeText', [Str] => Void );
    }
    ##############################################################################################
    #
    {
        my %seen;
        push @{ $EXPORT_TAGS{all} }, grep { !$seen{$_}++ } @{ $EXPORT_TAGS{$_} }
            for keys %EXPORT_TAGS;
        our @EXPORT_OK = @{ $EXPORT_TAGS{all} };
    }
};
1;
#
__END__

=pod

=encoding utf-8

=head1 NAME

LibUI - Simple, Portable, Native GUI Library

=head1 SYNOPSIS

    use LibUI qw[:all ::Window ::Label];
    Init( ) && die;
    my $window = LibUI::Window->new( 'Hi', 320, 100, 0 );
    $window->setMargined( 1 );
    $window->setChild( LibUI::Label->new('Hello, World!') );
    $window->onClosing(
        sub {
            Quit();
            return 1;
        },
        undef
    );
    $window->show;
    Main();

=begin html

<h2>Screenshots</h2> <div style="text-align: center"> <h3>Linux</h3><img
alt="Linux"
src="https://sankorobinson.com/LibUI.pm/screenshots/synopsis/linux.png" />
<h3>MacOS</h3><img alt="MacOS"
src="https://sankorobinson.com/LibUI.pm/screenshots/synopsis/macos.png" />
<h3>Windows</h3><img alt="Windows"
src="https://sankorobinson.com/LibUI.pm/screenshots/synopsis/windows.png" />
</div>

=end html

=head1 DESCRIPTION

LibUI is a simple and portable (but not inflexible) GUI library in C that uses
the native GUI technologies of each platform it supports.

This distribution is under construction. It works but is incomplete.

=head1 Container controls

=over

=item L<LibUI::Window> - a top-level window

=item L<LibUI::HBox> - a horizontally aligned, boxlike container that holds a group of controls

=item L<LibUI::VBox> - a vertically aligned, boxlike container that holds a group of controls

=item L<LibUI::Tab> - a multi-page control interface that displays one page at a time

=item L<LibUI::Group> - a container that adds a label to the child

=item L<LibUI::Form> - a container to organize controls as labeled fields

=item L<LibUI::Grid> - a container to arrange controls in a grid

=back

=head1 Data entry controls

=over

=item L<LibUI::Button> - a button control that triggers a callback when clicked

=item L<LibUI::Checkbox> - a user checkable box accompanied by a text label

=item L<LibUI::Entry> - a single line text entry field

=item L<LibUI::PasswordEntry> - a single line, obscured text entry field

=item L<LibUI::SearchEntry> - a single line search query field

=item L<LibUI::Spinbox> - display and modify integer values via a text field or +/- buttons

=item L<LibUI::Slider> - display and modify integer values via a draggable slider

=item L<LibUI::Combobox> - a drop down menu to select one of a predefined list of items

=item L<LibUI::EditableCombobox> - a drop down menu to select one of a predefined list of items or enter you own

=item L<LibUI::RadioButtons> - a multiple choice control of check buttons from which only one can be selected at a time

=item L<LibUI::DateTimePicker> - a control to enter a date and/or time

=item L<LibUI::DatePicker> - a control to enter a date

=item L<LibUI::TimePicker> - a control to enter a time

=item L<LibUI::MultilineEntry> - a multi line entry that visually wraps text when lines overflow

=item L<LibUI::NonWrappingMultilineEntry> - a multi line entry that scrolls text horizontally when lines overflow

=item L<LibUI::FontButton> - a control that opens a font picker when clicked

=back

=head2 Static controls

=over

=item L<LibUI::Label> - a control to display non-interactive text

=item L<LibUI::ProgressBar> - a control that visualizes the progress of a task via the fill level of a horizontal bar

=item L<LibUI::HSeparator> - a control to visually separate controls horizontally

=item L<LibUI::VSeparator> - a control to visually separate controls vertically

=back

=head2 Dialog windows

=over

=item C<openFile( )> - File chooser to select a single file

=item C<openFolder( )> - File chooser to select a single folder

=item C<saveFile( )> - Save file dialog

=item C<msgBox( ... )> - Message box dialog

=item C<msgBoxError( ... )> - Error message box dialog

=back

See L<LibUI::Window/Dialog windows> for more.

=head2 Menus

=over

=item C<LibUI::Menu> - application-level menu bar

=item C<LibUI::MenuItem> - menu items used in conjunction with L<LibUI::Menu>

=back

=head2 Tables

The upstream API is a mess so I'm still plotting around this.

=head1 GUI Functions

Some basics you gotta use just to keep a modern GUI running.

This is incomplete but... well, I'm working on it.

=head2 C<Init( [...] )>

    Init( );

Ask LibUI to do all the platform specific work to get up and running. If LibUI
fails to initialize itself, this will return a true value. Weird upstream
choice, I know...

You B<must> call this before creating widgets.

=head2 C<Main( ... )>

    Main( );

Let LibUI's event loop run until interrupted.

=head2 C<Uninit( ... )>

    Uninit( );

Ask LibUI to break everything down before quitting.

=head2 C<Quit( ... )>

    Quit( );

Quit.


=head2 C<Timer( ... )>

    Timer( 1000, sub { die 'do not do this here' }, undef);

    Timer(
        1000,
        sub {
            my $data = shift;
            return 1 unless ++$data->{ticks} == 5;
            0;
        },
        { ticks => 0 }
    );

Expected parameters include:

=over

=item C<$time>

Time in milliseconds.

=item C<$func>

CodeRef that will be triggered when C<$time> runs out.

Return a true value from your C<$func> to make your timer repeating.

=item C<$data>

Any userdata you feel like passing. It'll be handed off to your function.

=back

=head1 Importing Widgets

You're free to import widgets manually...

    use LibUI;
    use LibUI::Window;
    use LibUI::Button;
    use LibUI::VBox;
    use LibUI::Area;
    use LibUI::Entry;

...or, you can let LibUI take care of it:

    use LibUI qw[::Window ::Button ::VBox ::Area ::Entry];

Note the

=head1 Requirements

See L<Alien::libui>

=head1 See Also

F<eg/demo.pl> - Very basic example

F<eg/widgets.pl> - Demo of basic controls

=head1 LICENSE

Copyright (C) Sanko Robinson.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=head1 AUTHOR

Sanko Robinson E<lt>sanko@cpan.orgE<gt>

=for stopwords draggable gotta userdata

=cut

